File: workflow-ts/src/vision.ts
import { readFile } from "node:fs/promises";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import * as ort from "onnxruntime-node";
import sharp from "sharp";

export type ImageLabel = { label: string; score: number };

const ROOT = join(dirname(fileURLToPath(import.meta.url)), "..");
const MODEL_PATH = join(ROOT, "models", "squeezenet1.1-7.onnx");
const LABELS_PATH = join(ROOT, "imagenet-classes.txt");

const MEAN = [0.485, 0.456, 0.406];
const STD = [0.229, 0.224, 0.225];

let sessionPromise: Promise<ort.InferenceSession> | undefined;
let labelsPromise: Promise<string[]> | undefined;

function loadSession(): Promise<ort.InferenceSession> {
  sessionPromise ??= ort.InferenceSession.create(MODEL_PATH);
  return sessionPromise;
}

function loadLabels(): Promise<string[]> {
  labelsPromise ??= readFile(LABELS_PATH, "utf8").then((text) => text.split("\n").filter(Boolean));
  return labelsPromise;
}

async function preprocess(imageBase64: string): Promise<ort.Tensor> {
  const buffer = Buffer.from(imageBase64, "base64");
  const { data } = await sharp(buffer)
    .rotate()
    .removeAlpha()
    .resize(224, 224, { fit: "fill" })
    .raw()
    .toBuffer({ resolveWithObject: true });

  const float = new Float32Array(3 * 224 * 224);
  for (let i = 0; i < 224 * 224; i += 1) {
    const r = data[i * 3] / 255;
    const g = data[i * 3 + 1] / 255;
    const b = data[i * 3 + 2] / 255;
    float[i] = (r - MEAN[0]) / STD[0];
    float[224 * 224 + i] = (g - MEAN[1]) / STD[1];
    float[2 * 224 * 224 + i] = (b - MEAN[2]) / STD[2];
  }
  return new ort.Tensor("float32", float, [1, 3, 224, 224]);
}

/** Return the top ImageNet labels for a base64 JPEG/PNG. */
export async function labelImage(imageBase64: string, topK = 5): Promise<ImageLabel[]> {
  const [session, labels] = await Promise.all([loadSession(), loadLabels()]);
  const inputName = session.inputNames[0];
  const tensor = await preprocess(imageBase64);
  const result = await session.run({ [inputName]: tensor });
  const output = result[session.outputNames[0]];
  const logits = output.data as Float32Array;
  const max = Math.max(...logits);
  const exp = Array.from(logits, (value) => Math.exp(value - max));
  const sum = exp.reduce((acc, value) => acc + value, 0);
  const ranked = exp
    .map((value, index) => ({ index, score: value / sum }))
    .sort((a, b) => b.score - a.score)
    .slice(0, topK);

  return ranked.map(({ index, score }) => ({
    label: labels[index] ?? String(index),
    score,
  }));
}

File: workflow-py/vision.py
"""Turn a JPEG into ImageNet top labels with a local ONNX classifier."""

from __future__ import annotations

import base64
import io
from pathlib import Path

import numpy as np
import onnxruntime as ort
from PIL import Image

ROOT = Path(__file__).parent
MODEL_PATH = ROOT / "models" / "squeezenet1.1-7.onnx"
LABELS_PATH = ROOT / "imagenet_classes.txt"

_session: ort.InferenceSession | None = None
_labels: list[str] | None = None
_input_name: str | None = None

IMAGENET_MEAN = np.array([0.485, 0.456, 0.406], dtype=np.float32)
IMAGENET_STD = np.array([0.229, 0.224, 0.225], dtype=np.float32)


def _load() -> tuple[ort.InferenceSession, str, list[str]]:
    global _session, _labels, _input_name
    if _session is None:
        if not MODEL_PATH.exists():
            raise FileNotFoundError(
                f"missing classifier at {MODEL_PATH}; run python fetch_model.py"
            )
        _session = ort.InferenceSession(
            str(MODEL_PATH), providers=["CPUExecutionProvider"]
        )
        _input_name = _session.get_inputs()[0].name
    if _labels is None:
        _labels = LABELS_PATH.read_text(encoding="utf-8").splitlines()
    assert _input_name is not None
    return _session, _input_name, _labels


def _preprocess(image_base64: str) -> np.ndarray:
    raw = base64.b64decode(image_base64)
    image = Image.open(io.BytesIO(raw)).convert("RGB").resize((224, 224))
    array = np.asarray(image).astype(np.float32) / 255.0
    array = (array - IMAGENET_MEAN) / IMAGENET_STD
    return np.transpose(array, (2, 0, 1))[None, ...]


def label_image(image_base64: str, top_k: int = 5) -> list[dict[str, float | str]]:
    """Return the top ImageNet labels for a base64 JPEG/PNG."""
    session, input_name, labels = _load()
    tensor = _preprocess(image_base64)
    logits = session.run(None, {input_name: tensor})[0][0]
    logits = logits.astype(np.float64)
    logits = logits - logits.max()
    exp = np.exp(logits)
    probs = exp / exp.sum()
    order = np.argsort(probs)[::-1][:top_k]
    return [
        {"label": labels[int(i)] if int(i) < len(labels) else str(int(i)), "score": float(probs[int(i)])}
        for i in order
    ]
